src/darwin/Framework/CHIP/MTRDeviceConnectivityMonitor.mm

```diff
@@ -0,0 +1,247 @@
+        MTR_LOG_ERROR("%@ disconnected from dns-sd subsystem", self);
+        [self _stopMonitoring];
+        return;
+    }
+
+    // Create a nw_connection to monitor connectivity if the host name is not being monitored yet
+    NSString * hostNameString = [NSString stringWithUTF8String:hostName];
+    if (!_connections[hostNameString]) {
+        char portString[6];
+        snprintf(portString, sizeof(portString), "%d", ntohs(port));
+        nw_endpoint_t endpoint = nw_endpoint_create_host(hostName, portString);
+        if (!endpoint) {
+            MTR_LOG_ERROR("%@ failed to create endpoint for %s:%s", self, hostName, portString);
+            return;
+        }
+        nw_parameters_t params = nw_parameters_create_secure_udp(NW_PARAMETERS_DISABLE_PROTOCOL, NW_PARAMETERS_DEFAULT_CONFIGURATION);
+        if (!params) {
+            MTR_LOG_ERROR("%@ failed to create udp parameters", self);
+            return;
+        }
+        nw_connection_t connection = nw_connection_create(endpoint, params);
+        if (!connection) {
+            MTR_LOG_ERROR("%@ failed to create connection for %s:%s", self, hostName, portString);
+            return;
+        }
+        nw_connection_set_queue(connection, sSharedResolverQueue);
+        nw_connection_set_path_changed_handler(connection, ^(nw_path_t _Nonnull path) {
+            nw_path_status_t status = nw_path_get_status(path);
+            if (status == nw_path_status_satisfied) {
+                MTR_LOG_INFO("%@ path is satisfied", self);
+                std::lock_guard lock(sConnectivityMonitorLock);
+                [self _callHandler];
+            }
+        });
+        nw_connection_set_viability_changed_handler(connection, ^(bool viable) {
+            if (viable) {
+                std::lock_guard lock(sConnectivityMonitorLock);
+                MTR_LOG_INFO("%@ connectivity now viable", self);
+                [self _callHandler];
+            }
+        });
+        nw_connection_start(connection);
+    }
+}
+
+static void _resolveReplyCallback(
+    DNSServiceRef sdRef,
+    DNSServiceFlags flags,
+    uint32_t interfaceIndex,
+    DNSServiceErrorType errorCode,
+    const char * fullname,
+    const char * hosttarget,
+    uint16_t port, /* In network byte order */
+    uint16_t txtLen,
+    const unsigned char * txtRecord,
+    void * context)
+{
+    auto * connectivityMonitor = (__bridge MTRDeviceConnectivityMonitor *) context;
+    [connectivityMonitor handleResolvedHostname:hosttarget port:port error:errorCode];
+}
+
+- (void)startMonitoringWithHandler:(MTRDeviceConnectivityMonitorHandler)handler queue:(dispatch_queue_t)queue
+{
+    std::lock_guard lock(sConnectivityMonitorLock);
+
+    MTRDeviceConnectivityMonitorHandler handlerCopy = [handler copy];
+    _monitorHandler = handlerCopy;
+    _handlerQueue = queue;
+
+    // If there's already a resolver running, just return
+    if (_resolver) {
+        MTR_LOG_INFO("%@ connectivity monitor updated handler", self);
```

Comment:
The comment doesn't make sense. Should it be connectivity monitor already running or something like that or am I missing something here :)

Reply:
Maybe the intention was to update `_monitorHandler`? Or maybe this should just be a ```VerifyOrDie(!_resolver)```, since there is no expectation that there should be unbalanced `startMonitoring...` calls the way this class is currently used.